### ISSUE TITLE
Documentation Failure : Rbtree

### ISSUE THREAD

**Issue by wordjelly** [eventmachine/eventmachine#927]
Hi, 

I am hitting a very vague RBTREE Error, just running the code on your homepage, with em/pure_ruby.
Ruby Version : ruby 2.4.6p354 (2019-04-01 revision 67394) [x86_64-linux]
Eventmachine Version : 1.2.7

Running the code shown on the wiki, with "em/pure_ruby", instead of "eventmachine"

```
##test.rb
require "em/pure_ruby"

module EchoServer
   def post_init
     puts "-- someone connected to the echo server!"
   end

   def receive_data data
     send_data ">>>you sent: #{data}"
     close_connection if data =~ /quit/i
   end

   def unbind
     puts "-- someone disconnected from the echo server!"
   end
end

EventMachine.run {
  EventMachine.start_server "127.0.0.1", 8081, EchoServer
}
```

```
### Run the code
ruby test.rb
```

Error Trace:

```
/home/bhargav/.rbenv/versions/2.4.6/lib/ruby/2.4.0/set.rb:345:in `delete': can't modify rbtree during iteration (TypeError)
	from /home/bhargav/.rbenv/versions/2.4.6/lib/ruby/2.4.0/set.rb:345:in `delete'
	from /home/bhargav/.rbenv/versions/2.4.6/lib/ruby/gems/2.4.0/gems/eventmachine-1.2.7/lib/em/pure_ruby.rb:574:in `block in run_timers'
	from /home/bhargav/.rbenv/versions/2.4.6/lib/ruby/2.4.0/set.rb:324:in `each_key'
	from /home/bhargav/.rbenv/versions/2.4.6/lib/ruby/2.4.0/set.rb:324:in `each'
	from /home/bhargav/.rbenv/versions/2.4.6/lib/ruby/gems/2.4.0/gems/eventmachine-1.2.7/lib/em/pure_ruby.rb:569:in `run_timers'
	from /home/bhargav/.rbenv/versions/2.4.6/lib/ruby/gems/2.4.0/gems/eventmachine-1.2.7/lib/em/pure_ruby.rb:551:in `block in run'
	from /home/bhargav/.rbenv/versions/2.4.6/lib/ruby/gems/2.4.0/gems/eventmachine-1.2.7/lib/em/pure_ruby.rb:547:in `loop'
	from /home/bhargav/.rbenv/versions/2.4.6/lib/ruby/gems/2.4.0/gems/eventmachine-1.2.7/lib/em/pure_ruby.rb:547:in `run'
	from /home/bhargav/.rbenv/versions/2.4.6/lib/ruby/gems/2.4.0/gems/eventmachine-1.2.7/lib/em/pure_ruby.rb:144:in `run_machine'
	from /home/bhargav/.rbenv/versions/2.4.6/lib/ruby/gems/2.4.0/gems/eventmachine-1.2.7/lib/eventmachine.rb:195:in `run'
```

Any idea what's happening here?

Thanks.

**Comment by wordjelly** [comment]
I traced this problem to a function in the em/pure_ruby.rb file
```
module EventMachine
  class Reactor
    def run_timers
      timers_to_delete = []
      @timers.each {|t|
        if t.first <= @current_loop_time
          #this line throws the errors, as it attempts to delete an element of a sorted-set inside the iteration.
          ## => @timers.delete t
          timers_to_delete << t
          EventMachine::event_callback "", TimerFired, t.last
        else
          break
        end
      }
      timers_to_delete.map{|c| @timers.delete(c)}
    end
  end
end
```
The code above patches this, by deleting the timers after the iteration is complete.
AFAIK, this solves the problem.

I won't close the issue, as I am waiting to hear from one of the maintainers.

Thanks,
Bhargav

**Comment by sodabrew** [comment]
Good find! Generally, the pure-ruby implementation has not been as well-kept as it should be. I would figure this worked well enough in earlier versions of Ruby but was basically undefined behavior.

Definitely would take a PR that makes this change.

I was also thinking "why not use `delete_if`? but I believe you've done the right thing in the case of a large number of timers out of which only a subset need to be evaluated and executed. The short-circuit on the else case works because the timers are sorted, as you point out.

**Comment by wordjelly** [comment]
@sodabrew , thanks for the input, I'll get to it over the weekend.

**Comment by sodabrew** [comment]
Closed as Resolved by #929
